src/coreComponents/constitutive/fluid/NewMultiPhaseMultiComponentFluid.hpp

```diff
@@ -0,0 +1,870 @@
+                        arraySlice2d< real64 > const & dPhaseCompFraction_dPressure,
+                        arraySlice2d< real64 > const & dPhaseCompFraction_dTemperature,
+                        arraySlice3d< real64 > const & dPhaseCompFraction_dGlobalCompFraction,
+                        real64 & totalDensity,
+                        real64 & dTotalDensity_dPressure,
+                        real64 & dTotalDensity_dTemperature,
+                        arraySlice1d< real64 > const & dTotalDensity_dGlobalCompFraction ) const override;
+
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  virtual void update( localIndex const k,
+                       localIndex const q,
+                       real64 const pressure,
+                       real64 const temperature,
+                       arraySlice1d< real64 const > const & composition ) const override
+  {
+    compute( pressure,
+             temperature,
+             composition,
+             m_phaseFraction[k][q],
+             m_dPhaseFraction_dPressure[k][q],
+             m_dPhaseFraction_dTemperature[k][q],
+             m_dPhaseFraction_dGlobalCompFraction[k][q],
+             m_phaseDensity[k][q],
+             m_dPhaseDensity_dPressure[k][q],
+             m_dPhaseDensity_dTemperature[k][q],
+             m_dPhaseDensity_dGlobalCompFraction[k][q],
+             m_phaseMassDensity[k][q],
+             m_dPhaseMassDensity_dPressure[k][q],
+             m_dPhaseMassDensity_dTemperature[k][q],
+             m_dPhaseMassDensity_dGlobalCompFraction[k][q],
+             m_phaseViscosity[k][q],
+             m_dPhaseViscosity_dPressure[k][q],
+             m_dPhaseViscosity_dTemperature[k][q],
+             m_dPhaseViscosity_dGlobalCompFraction[k][q],
+             m_phaseCompFraction[k][q],
+             m_dPhaseCompFraction_dPressure[k][q],
+             m_dPhaseCompFraction_dTemperature[k][q],
+             m_dPhaseCompFraction_dGlobalCompFraction[k][q],
+             m_totalDensity[k][q],
+             m_dTotalDensity_dPressure[k][q],
+             m_dTotalDensity_dTemperature[k][q],
+             m_dTotalDensity_dGlobalCompFraction[k][q] );
+  }
+
+private:
+
+  /// Index of the gas phase
+  localIndex const m_phaseGasIndex;
+
+  /// Index of the liquid phase
+  localIndex const m_phaseLiquidIndex;
+
+  /// Kernel wrapper for brine viscosity updates
+  PVTProps::BrineViscosity::KernelWrapper const m_brineViscosityWrapper;
+
+  /// Kernel wrapper for CO2 viscosity updates
+  PVTProps::FenghourCO2Viscosity::KernelWrapper const m_co2ViscosityWrapper;
+
+  /// Kernel wrapper for brine density updates
+  PVTProps::BrineCO2Density::KernelWrapper const m_brineDensityWrapper;
+
+  /// Kernel wrapper for CO2 density updates
+  PVTProps::SpanWagnerCO2Density::KernelWrapper const m_co2DensityWrapper;
+
+  /// Kernel wrapper for phase fraction and phase component fraction updates
+  PVTProps::CO2Solubility::KernelWrapper const m_co2SolubilityWrapper;
```

Comment:
Hello @klevzoff,

On Lassen, I struggled with the data members in the update class of the new CO2-brine model (currently called `NewMultiPhaseMultiComponentFluid`):

1) In the current version of this PR, the data members are the wrappers of the PVTFunctions, such as, for instance, `PVTProps::BrineViscosity::KernelWrapper const m_brineViscosityWrapper;`. But, when I try to compile on Lassen, I get a compilation error saying:
```
Error: Formal parameter space overflowed
```
located where we launch the [fluid update](https://github.com/GEOSX/GEOSX/blob/11dffd4c2b3ad2e9716d86b53e5e8b929c36aaf5/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFlowKernels.hpp#L197) (unfortunately I forgot to save the exact error message, I do this in another branch when I remove the non-GPU callable fluid models from the `CMakeLists.txt`...). I found a [thread online](https://forums.developer.nvidia.com/t/formal-parameter-space-overflowed-4096-bytes-max/134513/3) saying that this error might be due to the way I pass the data to the GPU, so I tried 
something else (see point 2 below).

2) If I modify the update class to store views on the wrappers of the PVTFunctions:
```
 /// Kernel wrapper for brine viscosity updates                                                                                              
  arrayView1d< PVTProps::BrineViscosity::KernelWrapper const > const m_brineViscosityWrapper;                                                 
                                                                                                                                              
  /// Kernel wrapper for CO2 viscosity updates                                                                                                
  arrayView1d< PVTProps::FenghourCO2Viscosity::KernelWrapper const > const m_co2ViscosityWrapper;                                             
                                                                                                                                              
  /// Kernel wrapper for brine density updates                                                                                                
  arrayView1d< PVTProps::BrineCO2Density::KernelWrapper const > const m_brineDensityWrapper;                                                  
                                                                                                                                              
  /// Kernel wrapper for CO2 density updates                                                                                                  
  arrayView1d< PVTProps::SpanWagnerCO2Density::KernelWrapper const > const m_co2DensityWrapper;                                               
                                                                                                                                              
  /// Kernel wrapper for phase fraction and phase component fraction updates                                                                  
  arrayView1d< PVTProps::CO2Solubility::KernelWrapper const > const m_co2SolubilityWrapper;
```
then I can compile and run the CO2-brine simulations on Lassen, and everything seems to work fine.
However, this forces me to have these `arrayView1d`s of size 1 in the update class, which is a little odd. What do you think would be a good way to solve this issue?

Reply:
According to https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#function-parameters the total size of kernel arguments must not exceed 4 KB. We haven't hit this limit before, but it seems this object may indeed be quite large (if you add up the sizes of all `TableKernelWrapper`s and other members in its subobjects). Out of curiosity, can you print out `sizeof(NewMultiPhaseMultiComponentFluidUpdate)` somewhere?

An `ArrayView` of size 1 may be (much) smaller than its contained value, so I'm not surprised that it helps. This may be a reasonable workaround if it's not possible to reduce the size of the object. The only problem is that you also need to place the underlying `Array` object somewhere (inside `NewMultiPhaseMultiComponentFluid`?). I suppose we could also develop a host-device version of `unique_ptr` to provide a slightly lighter interface to manage a single object across memory spaces. 

@corbett5 @rrsettgast thoughts?

Reply:
Interesting. Yeah I'm not a huge fan of the `arrayView1D` containing a single object. A `unique_ptr` would be cleaner, but it wouldn't much easier to use than an array containing a single value. It would still need an owning class somewhere and a view class.

I wonder if captured variables count towards the parameter limit.

Reply:
> I wonder if captured variables count towards the parameter limit.

I think this is exactly what happens here. The captured-by-value object is a member of the capture object and contributes to its size. I wonder if RAJA could work around this issue on their side by dynamically allocating device memory for the device-side copy of the lambda, rather than passing it directly by-value through the CUDA launch API (conditionally, only when `sizeof(LABMDA)` above limit). Might be worth opening an issue with them.

Reply:
1) When I do `sizeof(NewMultiPhaseMultiComponentFluidUpdate)` before launching the kernel on the branch that has the size-1 `arrayView1d`s (see [here](https://github.com/GEOSX/GEOSX/blob/36311100493ee0a1f2a45acfb84e13ef2c0eeac4/src/coreComponents/constitutive/fluid/NewMultiPhaseMultiComponentFluid.hpp#L212-L225)), I get: 2.336 KB and it seems to work well.

2) When I try to compile on Lassen with the current version of this PR (no `arrayView1d` to store the `KernelWrapper`s), I get the following error message that, I think, contains the size of the kernel (5.920 KB):
```
/usr/WS1/GEOS/GEOSX/TPLs_2021-01-19/install-lassen-clang@upstream-release/raja/include/RAJA/policy/cuda/forall.hpp(137): 
Error: Formal parameter space overflowed (5920 bytes required, max 4096 bytes allowed) in function 
_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256ENS_9Iterators16numeric_iteratorIllPlEEZN5geosx34CompositionalMultiphaseFlowKernels17FluidUpdateKernel6launchINS1_9cuda_execILm256ELb0EEENS8_12constitutive38NewMultiPhaseMultiComponentFluidUpdateEEEvlRKT0_RKN7LvArray9ArrayViewIKdLi1ELi0ElNSJ_10ChaiBufferEEESP_dRKNSK_ISL_Li2ELi1ElSM_EEEUllE_lEEvT1_SG_T2_
```
And yes I forgot to say that for the version of 1) with the size-1 `arrayView1d`s I keep an `array1d` in `NewMultiPhaseMultiComponentFluid` [here](https://github.com/GEOSX/GEOSX/blob/36311100493ee0a1f2a45acfb84e13ef2c0eeac4/src/coreComponents/constitutive/fluid/NewMultiPhaseMultiComponentFluid.hpp#L335-L348).